README.md
# HawkScan

![alt tag](https://github.com/c0dejump/HawkScan/blob/master/static/logo.jpg)

Security Tool for Reconnaissance and Information Gathering on a website. (python 2.x & 3.x)

# News
**!** Add: Option --auth for HTTP authentification. Exemple --auth admin:admin      
**!** Fixed: Bugs     
*(for more details go on CHANGELOG.md)*

# Installation
 > 

       git clone https://github.com/c0dejump/HawkScan.git && sudo python HawkScan/setup.py

       pip(3) install -r requirements.txt 
    If problem with pip3:    
       sudo python3 -m pip install -r requirements.txt

 > 

# Special features
 - [x] Test backup/old file on all the files found (index.php.bak, index.php~ ...)
 - [x] Check header information
 - [x] Check DNS information
 - [x] Check whois information
 - [x] Check email in the website and in same time if emails leaked (report)
 - [x] CMS detection + version and vulns
 - [x] Backup system (if the script stopped, it take again in same place)
 - [x] WAF detection and Response error to WAF + Testing bypass it
 - [x] Check Github
 - [x] Option --exclude to exclude page, code error, bytes
 - [x] Work it with py2 and py3
 - [x] Option rate-limit if app is unstable (--timesleep)
 - [x] Check in waybackmachine
 - [x] Check if DataBase firebaseio existe and accessible
 - [x] Search S3 buckets in source code page
 - [x] Testing if it's possible scanning with "localhost" host
 - [x] Try differents bypass for 403 code error
 - [x] JS parsing and analysis (option --js)
 - [x] Check Google Dork 
 - [x] Check Host IP
 - [x] Auto resize relative to window
 
# TODO 
**P1 is the most important**

 - [ ] Multiple exclude like: --exclude 403,1337b [P1]
 - [ ] On-the-fly writing report [P2]
 - [ ] Option for forcing checking before scan start [P2]
 - [ ] Prefix filename (old_, copy of...) [P2]
 - [ ] Multiple website scanning [P2]
 - [ ] Check HTTP headers/ssl security [P2]
 - [ ] Anonymous routing through some proxy (http/s proxy list) [P2]
 - [ ] Access token [P2]
 - [ ] Check source code and verify leak or sensitive data in the Github [P2]
 - [ ] Analyse html code webpage [P3] => really necessary?
 - [ ] Check phpmyadmin version [P3]
 - [ ] Scan API endpoints/informations leaks [ASAP]
 - [ ] Active JS on website 2.0 (full js) + Webengine for MacOS [ASAP]
  
 >
     
    usage: hawkscan.py [-h] [-u URL] [-w WORDLIST] [-s SUBDOMAINS] [-t THREAD] [-a USER_AGENT] [--redirect] [-r] [-p PREFIX] [-o OUTPUT] [--cookie COOKIE_] [--exclude EXCLUDE] [--timesleep TS] [--auto] [--js]
 
 > 
 
    optional arguments: 
     -h, --help         show this help message and exit
     -u URL             URL to scan [required]
     -w WORDLIST        Wordlist used for URL Fuzzing. Default: dico.txt
     -s SUBDOMAINS      Subdomain tester
     -t THREAD          Number of threads to use for URL Fuzzing. Default: 20
     -a USER_AGENT      Choice user-agent 
     --redirect         For scan with redirect response (301/302) 
     -r                 Recursive dir/files      
     -p PREFIX          Add prefix in wordlist to scan      
     -o OUTPUT          Output to site_scan.txt (default in website directory)       
     -b                 Add a backup file scan like 'exemple.com/~exemple/, exemple.com/ex.php.bak...' but longer             
     -H HEADER_         modify HEADER              
     --exclude EXCLUDE  To define a page or response code status type to exclude during scan                                            
     --timesleep TS     To define a timesleep/rate-limit if app is unstable during scan                                 
     --auto             Automatic threads depending response to website. Max: 30      
     --update           For automatic update
     --js               For try to found keys or token in the javascript page  

 >

# Exemples

 >
    //Basic
     python hawkscan.py -u https://www.exemple.com/

    //With specific dico
     python hawkscan.py -u https://www.exemple.com/ -w dico_extra.txt

    //with 30 threads
     python hawkscan.py -u https://www.exemple.com/ -t 30

    //With backup files scan
     python hawkscan.py -u https://www.exemple.com/ -b

    //With an exclude page
     python hawkscan.py -u https://www.exemple.com/ --exclude https://www.exemple.com/profile.php?id=1

    //With an exclude response code
     python hawkscan.py -u https://www.exemple.com/ --exclude 403

    //With an exclude bytes number
     python hawkscan.py -u https://www.exemple.com/ --exclude 1337b 

 >

# Thanks
Layno (https://github.com/Clayno/) [Technical helper]      
Sanguinarius (https://twitter.com/sanguinarius_Bt) [Technical helper]         
Cyber_Ph4ntoM (https://twitter.com/__PH4NTOM__) [Beta tester]


# Paypal

https://www.paypal.me/c0dejump

## Tools used

This script use "WafW00f" to detect the WAF in the first step (https://github.com/EnableSecurity/wafw00f)

This script use "Sublist3r" to scan subdomains (https://github.com/aboul3la/Sublist3r)

This script use "waybacktool" to check in waybackmachine (https://github.com/Rhynorater/waybacktool)

This script use "degoogle" to check google dork queries (https://github.com/deepseagirl/degoogle)


CHANGELOG.md
Changelog:
----------

- 1.6.7
---------
	Add: Option --auth for HTTP authentification. Exemple --auth admin:admin
	Fixed: Bugs
---------

- 1.6.6
---------
	Faster
	Updated: Less FP with bypass forbidden function
	Updated: Less errors in princpal script
	Updated: Little modifications in report
	Fixed: Style and system bugs
	Added: Auto resize relative to window
---------

- 1.6
---------
	Added: "degoogle" tools for google dork queries, more fast and less of google captcha
	Updated: Code optimization
---------

- 1.5.9.1
---------
	Fixed: System bugs
---------

- 1.5.9
---------
	Started: Code optimization
	Updated: Changed changelog.md + Readme.md
---------

- 1.5.8
---------
	Updated: file and directory function management
			A directory of the website is automatically create like: "website_date/".	
			And if the directory exist an other directory is created like: "website date_hour/"
---------

- 1.5.7
---------
	Added: Function to try if the website IP is different of the website domain
	Updated: dico.txt
---------
	
- 1.5.6
---------
	Fixed: system bugs
	Added: New exclude type, now you can exclude a number of byte with "b" at the end in your number, like: --exclude 550b
---------
	
- 1.5.5
---------
	Added: Google dork requests at the first scan
---------
	
- 1.5.4
---------
	Added: Option "--js" for scan and analyse JS
	Deleted: "dryscrape" librarie for the moment, many error with it, I'll remake it later
---------
	
- 1.5.3
---------
	Added: Setup.py, you can just doing "python setup.py"
---------
	
- 1.5.2
---------
	Added: Try differents bypass for 403 code error
	Updated: dico.txt
---------
	
- 1.5.1
---------
	New banner
	Fix bugs
---------
	
- 1.5
---------
	~~Auto activate JS during scan if the webite is full JS (website 2.0)~~
---------
	
- 1.4
---------
	Add: Dockerfile
---------
	
- 1.3.3
---------
	Add: New function which try automatically if it's possible scanning with "localhost" host
---------
	
- 1.3.2
---------
	Replace: "--cookie" by "-H" for different header values; ex: -H "Host:test" // -H "Authentification:cookie" (not space after ":" or "=")
---------
	
- 1.3.1
---------
	Code review
	New logo
	Adding Changelog
---------
	
- 1.2
---------
	Adding news words in dico.txt (old dico_extra.txt)
	Adding extensions in backup check test function, option -b (.json, .xml, .bkp...) => very long
	Test bypass of waf rate limited in real time (X-Originating-IP...)
	Exclude response http code (--exclude 403)
	Filter on response http code in report
---------
	
- 1.0
---------
  	Better management Threads
	Add news words in dico_extra.txt
	New style for the report
	Errors log management
---------

modules/before_run.py
import requests
from bs4 import BeautifulSoup
import whois
import json
import sys, re, os
import ssl, OpenSSL
import socket
import pprint
import traceback
from requests.exceptions import Timeout
import time
# External
from config import PLUS, WARNING, INFO, LESS, LINE, FORBI, BACK

class before_start:
    

    def get_header(self, url, directory):
        """Get header of website (cookie, link, etc...)"""
        r = requests.get(url, allow_redirects=False, verify=False)
        print(INFO + "HEADER")
        print(LINE)
        print("  {} \n".format(r.headers).replace(',','\n'))
        print(LINE)
        with open(directory + '/header.csv', 'w+') as file:
            file.write(str(r.headers).replace(',','\n'))


    def gitpast(self, url):
        """
        Github: check github informations
        Pastebin: check pastebin information #TODO
        """
        print("{}Check in Github".format(INFO))
        print(LINE)
        url = url.split(".")[1] if "www" in url else url.split("/")[2]
        url = "{}".format(url)
        print("search: {}\n".format(url))
        types = ["Commits", "Issues", "Code", "Repositories", "Marketplace", "Topics", "Wikis", "Users"]
        for t in types:
            github = "https://github.com/search?q={}&type={}".format(url, t)
            req = requests.get(github, verify=False)
            soup = BeautifulSoup(req.text, "html.parser")
            search = soup.find('a', {"class":"menu-item selected"})
            if search:
                for s in search.find("span"):
                    print("  {}{}: {}".format(INFO, t, s))
            else:
                print("  {}{}: not found".format(INFO, t))
        print("\n" + LINE)


    def who_is(self, url, directory):
        """Get whois of website"""
        print(INFO + "WHOIS")
        print(LINE)
        try:
            who_is = whois.whois(url)
            #pprint.pprint(who_is + "\n")
            for k, w in who_is.iteritems():
                is_who = " {} : {}-".format(k, w)
                print(is_who)
                with open(directory + '/whois.csv', 'a+') as file:
                    file.write(is_who.replace("-","\n"))
        except:
            erreur = sys.exc_info()
            typerr = u"%s" % (erreur[0])
            typerr = typerr[typerr.find("'")+1:typerr.rfind("'")]
            print(typerr)
            msgerr = u"%s" % (erreur[1])
            print(msgerr)
        print("\n" + LINE)


    def get_dns(self, url, directory):
        """Get DNS informations"""
        port = 0
        try:
            if "https" in url:
                url = url.replace('https://','').replace('/','')
                port = 443
            else:
                url = url.replace('http://','').replace('/','')
                port = 80
            context = ssl.create_default_context()
            conn = context.wrap_socket(socket.socket(socket.AF_INET), server_hostname=url)
            conn.connect((url, port))
            cert = conn.getpeercert()
            print(INFO + "DNS information")
            print(LINE)
            pprint.pprint(str(cert['subject']).replace(',','').replace('((','').replace('))',''))
            pprint.pprint(cert['subjectAltName'])
            print('')
            conn.close()
            print(LINE)
            with open(directory + '/dns_info.csv', 'w+') as file:
                file.write(str(cert).replace(',','\n').replace('((','').replace('))',''))
        except:
            print(INFO + "DNS information")
            print(LINE)
            erreur = sys.exc_info()
            typerr = u"%s" % (erreur[0])
            typerr = typerr[typerr.find("'")+1:typerr.rfind("'")]
            print(typerr)
            msgerr = u"%s" % (erreur[1])
            print(msgerr + "\n")
            print(LINE)


    def firebaseio(self, url):
        """
        Firebaseio: To check db firebaseio
        ex: --firebase facebook
        """
        get_domain = url.split("/")[2]
        parse_domain = get_domain.split(".")
        if not "www" in get_domain:
            dire = "{}-{}".format(parse_domain[0], parse_domain[1]) if len(parse_domain) > 2 else "{}".format(parse_domain[0])
        else:
            dire = "{}".format(parse_domain[1])
        print("{}Firebaseio Check".format(INFO))
        print(LINE)
        url = 'https://{}.firebaseio.com/.json'.format(dire.split(".")[0])
        print(url + "\n")
        r = requests.get(url, verify=False).json()
        try:
            if 'error' in r.keys():
                if r['error'] == 'Permission denied':
                    print("\t{}{} seems to be protected".format(FORBI, url)) #successfully protected
                elif r['error'] == '404 Not Found':
                    print("\t{}{} not found".format(LESS, url)) #doesn't exist
                elif "Firebase error." in r['error']:
                    print("\t{}{} Firebase error. Please ensure that you spelled the name of your Firebase correctly ".format(WARNING, url))
            else:
                print("\t{}{} seems to be vulnerable !".format(PLUS, url)) #vulnerable
        except AttributeError:
            '''
            Some DBs may just return null
            '''
            print("{} null return".format(INFO))
        print(LINE)
        

    def wayback_check(self, url, directory):
        """
        Wayback_check:
        Check in a wayback machine to found old file on the website or other things...
        Use "waybacktool"
        """
        print("{}Wayback Check".format(INFO))
        print(LINE)
        print(url + "\n")
        try:
            os.system('python tools/waybacktool/waybacktool.py pull --host {} | python tools/waybacktool/waybacktool.py check > {}/wayback.txt'.format(url, directory))
        except Exception:
            traceback.print_exc()
        try:
            statinfo = os.path.getsize(directory + "/wayback.txt")
        except:
            print("\t{}Nothing wayback found !".format(INFO))
        if statinfo < 1 :
            print("\t{}Nothing wayback found".format(INFO))
        else:
            with open(directory + "/wayback.txt", "r+") as wayback:
                wb_read = wayback.read().splitlines()
                for wb in wb_read:
                    wb_res = list(wb.split(","))
                    try:
                        if wb_res[1] == " 200":
                            print("{}{}{}".format(PLUS, wb_res[0], wb_res[1]))
                        elif wb_res[1] == " 301" or wb_res[1] == " 302":
                            print("{}{}{}".format(LESS, wb_res[0], wb_res[1]))
                        elif wb_res[1] == " 404" or wb_res[1] == " 403":
                            pass
                        else:
                            print("{}{}{}".format(INFO, wb_res[0], wb_res[1]))
                    except:
                        pass
        print(LINE)


    def check_localhost(self, url):
        #TODO
        """
        CHeck_localhost: Function which try automatically if it's possible scanning with "localhost" host for discovery other files/directories
        """
        list_test = ["127.0.0.1", "localhost"]
        localhost = False
        print("{}Try localhost host".format(INFO))
        print(LINE)
        for lt in list_test:
            header = {"Host": lt}
            try:
                req = requests.get(url, headers=header, verify=False, timeout=10)
                if req.status_code == 200:
                    print("\t{}You can potentialy try bf directories with this option '-H \"Host:{}\"' ".format(PLUS, lt))
                    localhost = True
                else:
                    pass
            except:
                pass
        if not localhost:
            print("\t{}Not seem possible to scan with localhost host".format(LESS))
        print(LINE)


    def check_ip(self, domain, url):
        """
        check_ip:
        Check the host ip if this webpage is different or not
        """
        print("{}Check Host IP".format(INFO))
        print(LINE)
        req_index = requests.get(url, verify=False)
        len_index = len(req_index.content)
        retrieve_ip = False
        dom = socket.gethostbyname(domain)
        ips = ["https://{}/".format(dom), "http://{}/".format(dom)]
        for ip in ips:
            try:
                req_ip = requests.get(ip, verify=False)
                if req_ip.status_code not in [404, 403, 425, 503, 500, 400] and len(req_ip.content) != len_index:
                    retrieve_ip = True
                    print("\t{}The host IP seem to be different, check it: {} ".format(PLUS, ip))
            except:
                pass
        if not retrieve_ip:
            print("\t{}The IP Not seem different host".format(LESS))
        print(LINE)


    def test_timeout(self, url):
        """
        Test_timeout: just a little function for test if the connection is good or not
        """
        try:
            req_timeout = requests.get(url, timeout=30)
        except Timeout:
            print("{}Service potentialy Unavailable, The site web seem unavailable please wait...\n".format(WARNING))
            time.sleep(180)
        except requests.exceptions.ConnectionError:
            pass